### PR TITLE
feat: Nat.sqrt

### PR DESCRIPTION
**Pull request by JamesGallicchio**
Unclear whether this should be an extern'd function in Lean core. But this implementation has fine enough performance for now.

### CODE REVIEW

File: Std/Data/Nat/Basic.lean [hunk 1]
```diff
@@ -15,3 +15,19 @@ instance : Dvd Nat := ⟨fun a b => ∃ c, b = a * c⟩
 
 /-- Sum of a list of natural numbers. -/
 protected def sum (l : List Nat) : Nat := l.foldr (·+·) 0
+
+/--
+Integer square root function. Implemented via Newton's method.
+-/
+def sqrt (n : Nat) : Nat :=
+  let guess := n / 2
+  if guess = 0 then n else
```
Comment (digama0) [inline Std/Data/Nat/Basic.lean:24]:
```suggestion
  if n ≤ 1 then n else
  let guess := 1 <<< ((n - 1).log2 >>> 1 + 1)
```

File: Std/Data/Nat/Basic.lean [hunk 2]
```diff
@@ -15,3 +15,19 @@ instance : Dvd Nat := ⟨fun a b => ∃ c, b = a * c⟩
...
+  let rec iter (guess : Nat) : Nat :=
+    let next := (guess + n / guess) / 2
```
Comment (digama0) [inline Std/Data/Nat/Basic.lean:26]:
```suggestion
    let next := (guess + n / guess) >>> 1
```
Would be nice if the compiler automatically did strength reduction but it doesn't seem to looking at the IR.

File: Std/Data/Nat/Basic.lean [hunk 3]
```diff
@@ -15,3 +15,19 @@ instance : Dvd Nat := ⟨fun a b => ∃ c, b = a * c⟩
...
+    if h : guess ≤ next then
+      guess
+    else
+      have : next < guess := Nat.gt_of_not_le h
+      iter next
```
Comment (digama0) [inline Std/Data/Nat/Basic.lean:31]:
```suggestion
    if h : next < guess then iter next else guess
```

File: Std/Data/Nat/Basic.lean (hunk 2 above)
Comment (hargoniX) [reply Std/Data/Nat/Basic.lean:26]: I'll add that to the new code generator when the LetExpr refactoring is done.

**Comment by digama0** [comment]
On second thought, I think the sqrt lemmas should be put in a separate PR. It was almost ready to merge before but now it looks like a bit more work.

**Pull request closed without merging**